tests/integration_tests/performance/test_vsock_throughput.py

```diff
@@ -0,0 +1,381 @@
+                                           # Distribute the modes evenly.
+                                           modes[client_idx % modes_len]))
+
+        cpu_load = cpu_load_future.result()
+        for future in futures[:-1]:
+            res = json.loads(future.result())
+            res[IPERF3_END_RESULTS_TAG][
+                IPERF3_CPU_UTILIZATION_PERCENT_OUT_TAG] = None
+            yield res
+
+        # Attach the real CPU utilization vmm/vcpus to
+        # the last iperf3 server-client pair measurements.
+        res = json.loads(futures[-1].result())
+
+        # We expect a single emulation thread tagged with `firecracker` name.
+        tag = "firecracker"
+        assert tag in cpu_load and len(cpu_load[tag]) == 1
+        thread_id = list(cpu_load[tag])[0]
+        data = cpu_load[tag][thread_id]
+        vmm_util = sum(data)/len(data)
+        cpu_util_perc = res[IPERF3_END_RESULTS_TAG][
+            IPERF3_CPU_UTILIZATION_PERCENT_OUT_TAG] = dict()
+        cpu_util_perc[CPU_UTILIZATION_VMM_TAG] = vmm_util
+
+        vcpus_util = 0
+        for vcpu in range(basevm.vcpus_count):
+            # We expect a single fc_vcpu thread tagged with
+            # f`fc_vcpu {vcpu}`.
+            tag = f"fc_vcpu {vcpu}"
+            assert tag in cpu_load and len(cpu_load[tag]) == 1
+            thread_id = list(cpu_load[tag])[0]
+            data = cpu_load[tag][thread_id]
+            vcpus_util += (sum(data)/len(data))
+
+        cpu_util_perc[CPU_UTILIZATION_VCPUS_TOTAL_TAG] = vcpus_util
+
+        yield res
+
+
+def consume_iperf_output(cons,
+                         result,
+                         vcpus_count):
+    """Consume iperf3 output result for TCP workload."""
+    total_received = result[IPERF3_END_RESULTS_TAG]['sum_received']
+    duration = float(total_received['seconds'])
+    cons.consume_stat(DURATION_TOTAL, DURATION, duration)
+
+    # Computed at the receiving end.
+    total_recv_bytes = int(total_received['bytes'])
+    tput = round((total_recv_bytes*8) / (1024*1024*duration), 2)
+    cons.consume_stat(THROUGHPUT_TOTAL, THROUGHPUT, tput)
+
+    cpu_util = result[IPERF3_END_RESULTS_TAG][
+        IPERF3_CPU_UTILIZATION_PERCENT_OUT_TAG]
+    if cpu_util:
+        cpu_util_host = cpu_util[CPU_UTILIZATION_VMM_TAG]
+        cpu_util_guest = cpu_util[CPU_UTILIZATION_VCPUS_TOTAL_TAG]
+
+        cons.consume_stat(CPU_UTILIZATION_VMM_TAG,
+                          CPU_UTILIZATION_HOST,
+                          cpu_util_host)
+        cons.consume_stat(CPU_UTILIZATION_VCPUS_TOTAL_TAG,
+                          CPU_UTILIZATION_GUEST,
+                          cpu_util_guest)
+
+
+def pipes(basevm, current_avail_cpu, env_id):
+    """Producer/Consumer pipes generator."""
+    for mode in CONFIG["modes"]:
+        # We run bi-directional tests only on uVM with more than 2 vCPus
+        # because we need to pin one iperf3/direction per vCPU, and since we
+        # have two directions, we need at least two vCPUs.
```

Comment:
Why do we need at least two vCPUs? We can still run a bidirectional test and the iperf will share a CPU.

Reply:
As the comment says, because of CPU pinning. We could pin 2 clients on the same Vcpu, but this will create contention and results with lower reliability (depending on the CPU scheduler's decisions)